### PR TITLE
AccName: additional text node tests

### PR DESCRIPTION
**Pull request by cookiecrook**
Closes https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/37

### CODE REVIEW

**Comment by cookiecrook** [comment]
@sivakusayan @aleventhal @MelSumner @jnurthen @jcsteh

Do you agree with my test expectations about interior whitespace?

For example, several tests are failing across the board (see CI results in the checks above: [WebKit](https://wpt.fyi/results/accname/name/comp_text_node.html?diff&filter=ADC&run_id=6311522617262080&run_id=5185622710419456), [Chromium](https://wpt.fyi/results/accname/name/comp_text_node.html?diff&filter=ADC&run_id=5097079677911040&run_id=5192894022942720), [Gecko](https://wpt.fyi/results/accname/name/comp_text_node.html?diff&filter=ADC&run_id=5184921972244480&run_id=5129020074622976)) b/c of a distinction between regular space chars and non-breaking space chars. It's interior space, which is less strictly specified than leading/trailing space.
```
!EQ("button label", "button label")
```

### Of Note

The WPT helper method `verifyLabelsBySelector()` is explicitly trimming leading/trailing whitespace b/c in most cases, it doesn't matter. 
https://github.com/web-platform-tests/wpt/blob/master/wai-aria/scripts/aria-utils.js#L141

We could also normalize interior whitespace to match, and these tests would pass across the board, and add a new method that preserves whitespace exactly, which we could use on a limited number of tests.

**Comment by MelSumner** [comment]
> We could also normalize interior whitespace to match (these tests would then pass across the board) and add a new method that preserves whitespace exactly, which we could use on a limited number of tests.

This is my preference FWIW. Maybe with a comment in there to revisit in the future. I think there are some whitespace issues that need to be resolved just because we can't seem to get consensus on where they should be  or where they should be compressed due to different interpretations and implementations.

**Comment by cookiecrook** [comment]
Seems like that’s the consensus from the related AccName issue linked above, so I’ll make it happen. Thanks all.

**Comment by cookiecrook** [comment]
Yuck. I'll fix it.

File: wai-aria/scripts/aria-utils.js [hunk 1]
```diff
@@ -136,10 +136,17 @@ const AriaUtils = {
...
+        let expectedLabel = el.getAttribute("data-expectedlabel");
         let computedLabel = await test_driver.get_computed_label(el);
-        // Todo: Remove whitespace normalization after https://github.com/w3c/accname/issues/192 is addressed. Change prior line back to `const`, too.
-        computedLabel = computedLabel.trim()
+
+        // See: https://github.com/w3c/accname/issues/192 and https://github.com/w3c/accname/issues/208
+        //
+        // HTML/JS whitespace matches regular space, formfeed (\f), newline (\n), return (\r), tab (\t), vertical tab (\v),
+        // non-breaking space (\xA0), and a few more Unicode whitespace characters: \u00A0, \u2028, and \u2029.
+        const whitespace = /\s+/g;
```
Comment (zcorpan) [inline wai-aria/scripts/aria-utils.js:146]: https://github.com/w3c/accname/pull/165 resolved to use https://infra.spec.whatwg.org/#ascii-whitespace (but seemingly only for the `aria-label` attribute). `\s` matches more characters: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions/Cheatsheet#character_classes

Comment (cookiecrook) [reply wai-aria/scripts/aria-utils.js]: Thanks. Updated. 
https://github.com/web-platform-tests/wpt/pull/42407/files

File: accname/name/comp_text_node.html [hunk 2]
```diff
@@ -13,23 +13,96 @@
...
+<br>
+
+<h1>text node, with leading/trailing non-breaking space</h1>
+<button class="ex" data-expectedlabel="button&nbsp;label" data-testname="button with text node, with leading/trailing non-breaking space">&nbsp;button&nbsp;label&nbsp;</button>
+<h3 class="ex" data-expectedlabel="heading&nbsp;label" data-testname="heading with text node, with leading/trailing non-breaking space">&nbsp;heading&nbsp;label&nbsp;</h3>
+<a href="#" class="ex" data-expectedlabel="link&nbsp;label" data-testname="link with text node, with leading/trailing non-breaking space">&nbsp;link&nbsp;label&nbsp;</a>
+<br>
+
+<h1>text node, with leading/trailing non-breaking space</h1>
+<button class="ex" data-expectedlabel="button &nbsp; label" data-testname="button with text node, with mixed space and non-breaking space">button   &nbsp;   label</button>
```
Comment (sivakusayan) [inline accname/name/comp_text_node.html:81]: Thanks for writing all these tests!

I wasn't 100% sure how regular whitespace should be interacting with non-breaking spaces when I made the [Blink PR](https://github.com/chromium/chromium/commit/68243ce9b062184744635e3d75a39d429845ee7e), but it seems for now it should conform with the expectations you had here.

I was curious whether we should collapse `&nbsp;` and regular U+0020 spaces together into a single U+0020 space, but I wasn't able to find much guidance. I tried to look to the WHATWG Infra standard that seems to [define what it means to "collapse ASCII whitespace"](https://infra.spec.whatwg.org/#strip-and-collapse-ascii-whitespace),  but interestingly non-breaking spaces aren't included in the definition of ASCII whitespace in that spec. So I decided to leave that area alone for now.

Comment (cookiecrook) [reply accname/name/comp_text_node.html:81]: Yes, I'm thankful @zcorpan pointed that out earlier in the review, and I've updated to match the HTML definition of "ASCII Whitespace" that ARIA/AccName references.

Comment (cookiecrook) [reply accname/name/comp_text_node.html:81]: More discussion in [AccName #208](https://github.com/w3c/accname/issues/208).

File: wai-aria/scripts/aria-utils.js [hunk 3]
```diff
@@ -136,10 +136,22 @@ const AriaUtils = {
       let label = el.getAttribute("data-expectedlabel");
       let testName = el.getAttribute("data-testname") || label; // data-testname optional if label is unique per test file
       promise_test(async t => {
-        const expectedLabel = el.getAttribute("data-expectedlabel");
+        let expectedLabel = el.getAttribute("data-expectedlabel");
```
Comment (cookiecrook) [inline wai-aria/scripts/aria-utils.js:139]:
```suggestion
        const expectedLabel = el.getAttribute("data-expectedlabel");
```

File: wai-aria/scripts/aria-utils.js [hunk 4]
```diff
@@ -136,10 +136,22 @@ const AriaUtils = {
...
+        // - https://github.com/w3c/accname/pull/165
+        // - https://github.com/w3c/accname/issues/192
+        // - https://github.com/w3c/accname/issues/208
+        //
+        // AccName reference's HTML's definition of ASCII Whitespace
+        // https://infra.spec.whatwg.org/#ascii-whitespace
+        // which matches tab (\t), newline (\n), formfeed (\f), return (\r), and regular space (\u0020).
+        // but it does NOT match non-breaking space (\xA0,\u00A0) and others matched by \s
+        const asciiWhitespace = /\t\n\f\r\u0020+/g;
+        expectedLabel = expectedLabel.replace(asciiWhitespace, '\u0020').trim();
```
Comment (cookiecrook) [inline wai-aria/scripts/aria-utils.js:152]:
```suggestion
```

Comment (cookiecrook) [reply wai-aria/scripts/aria-utils.js:152]: In hindsight, it seems obvious that I should not be normalizing trimming the expectation, too! 🤣

File: wai-aria/scripts/aria-utils.js [hunk 5]
```diff
@@ -138,8 +138,19 @@ const AriaUtils = {
...
         let computedLabel = await test_driver.get_computed_label(el);
-        // Todo: Remove whitespace normalization after https://github.com/w3c/accname/issues/192 is addressed. Change prior line back to `const`, too.
-        computedLabel = computedLabel.trim()
+
+        // See:
+        // - https://github.com/w3c/accname/pull/165
+        // - https://github.com/w3c/accname/issues/192
+        // - https://github.com/w3c/accname/issues/208
+        //
+        // AccName reference's HTML's definition of ASCII Whitespace
```
Comment (cookiecrook) [inline wai-aria/scripts/aria-utils.js:147]:
```suggestion
        // AccName references HTML's definition of ASCII Whitespace
```

**Comment by cookiecrook** [comment]
Recycling CI.

File: wai-aria/scripts/aria-utils.js [hunk 6]
```diff
@@ -138,8 +138,19 @@ const AriaUtils = {
...
+        // - https://github.com/w3c/accname/pull/165
+        // - https://github.com/w3c/accname/issues/192
+        // - https://github.com/w3c/accname/issues/208
+        //
+        // AccName references HTML's definition of ASCII Whitespace
+        // https://infra.spec.whatwg.org/#ascii-whitespace
+        // which matches tab (\t), newline (\n), formfeed (\f), return (\r), and regular space (\u0020).
+        // but it does NOT match non-breaking space (\xA0,\u00A0) and others matched by \s
+        const asciiWhitespace = /\t\n\f\r\u0020+/g;
+        computedLabel = computedLabel.replace(asciiWhitespace, '\u0020').trim();
```
Comment (jugglinmike) [inline wai-aria/scripts/aria-utils.js:152]: `trim` operates on [the same characters](https://tc39.es/ecma262/#sec-trimstring) as [RegExp's `\u` character class](https://tc39.es/ecma262/#sec-compiletocharset).

```suggestion
        computedLabel = computedLabel.replace(asciiWhitespace, '\u0020')
          .replace(/^\u0020|\u0020$/g, '');
```

Comment (jugglinmike) [reply wai-aria/scripts/aria-utils.js:152]: Rather, RegExp's `\s` character class.

Comment (cookiecrook) [reply wai-aria/scripts/aria-utils.js:152]:
```suggestion
        computedLabel = computedLabel.replace(asciiWhitespace, '\u0020').replace(/^\u0020|\u0020$/g, '');
```

**Review by jugglinmike — COMMENTED** [review]
Sorry I missed this on my first review!

File: accname/name/comp_text_node.html [hunk 7]
```diff
@@ -13,23 +13,96 @@
 
 <p>Tests the <a href="https://w3c.github.io/accname/#comp_text_node">#comp_text_node</a> portions of the AccName <em>Name Computation</em> algorithm.</p>
```
Comment (rahimabdi) [inline accname/name/comp_text_node.html:14]: Having trouble nailing down what constitutes a "text node"; in the spec, it is capitalized as "Text" so perhaps it refers to the DOM "Text" interface? https://dom.spec.whatwg.org/#interface-text

I understood this step of the comp algorithm as testing nodes that are non-interactive and only contain text (e.g., headings, paragraph, caption). Do we need greater clarification on what a "text node" is?

File: accname/name/comp_text_node.html [hunk 8]
```diff
@@ -13,23 +13,96 @@
...
-<!-- I'm not certain whether #comp_text_node requires a lot of testing outside of the #comp_name_from_content contexts, -->
-<!-- but I did think of one example where text node versus comment node may make a difference when joining text nodes with a space vs innerText. -->
+<!-- Note: some overlap with the tests in ./comp_name_from_content.html -->
 
-<!-- Skipped (class="ex" removed) until https://github.com/w3c/accname/issues/193 is resolved -->
-<h2 class="ex-skipped" data-expectedlabel="heading label" data-testname="heading with text/comment/text nodes, no space">
-  heading<!-- with non-text node splitting concatenated text nodes -->label<!-- [sic] no extra spaces around first comment -->
-</h2>
 
+<h1>text/element/text nodes, no space</h1>
```
Comment (rahimabdi) [inline accname/name/comp_text_node.html:19]: Minor suggestion on usage of multiple <h1>s (although unsure if that was intentional).

Comment (rahimabdi) [reply accname/name/comp_text_node.html:19]: Also curious if there are children text nodes that  are both empty and non-empty, e.g.,:

```
<button>
     button
     <span>label</span>
     <span></span>
     <span>foo</span>
</button>
```

(`computedLabel = buttonlabelfoo`)

File: accname/name/comp_text_node.html [hunk 9]
```diff
@@ -13,23 +13,96 @@
 
 <p>Tests the <a href="https://w3c.github.io/accname/#comp_text_node">#comp_text_node</a> portions of the AccName <em>Name Computation</em> algorithm.</p>
 
-<!-- I'm not certain whether #comp_text_node requires a lot of testing outside of the #comp_name_from_content contexts, -->
-<!-- but I did think of one example where text node versus comment node may make a difference when joining text nodes with a space vs innerText. -->
+<!-- Note: some overlap with the tests in ./comp_name_from_content.html -->
```
Comment (rahimabdi) [inline accname/name/comp_text_node.html:16]: Additional tests I would propose:
- Text nodes as part of explicit ARIA roles
- Tests that ensure labelledby/aria-label/name from content etc. take precedence over text nodes (may be captured in respective files)

File: accname/name/comp_text_node.html (hunk 7 above)
Comment (cookiecrook) [reply accname/name/comp_text_node.html:14]: These are inherited from XML terms: most commonly _text node_ versus _element node_ versus _comment node_. Element nodes can contain other element, comment, and text nodes, but IIRC text nodes are always leaf nodes.

Comment (cookiecrook) [reply accname/name/comp_text_node.html:14]: > I initially understood this step of the comp algorithm as testing nodes that are non-interactive and only contain text (e.g., headings, paragraph, caption), or text as children nodes of HTML tags. 

hence testing the text nodes split by comment nodes, with or without additional spaces

```html
<h3 class="ex" data-expectedlabel="headinglabel" data-testname="heading with text/comment/text nodes, no space">
  heading<!-- with non-text node splitting concatenated text nodes -->label<!-- [sic] no extra spaces around first comment -->
</h3>
```

Comment (cookiecrook) [reply accname/name/comp_text_node.html:14]: Maybe your comment is that AccName should link to specific portion of the DOM spec? If so, +1

File: accname/name/comp_text_node.html (hunk 8 above)
Comment (cookiecrook) [reply accname/name/comp_text_node.html:19]: > Minor suggestion on usage of multiple `h1`s (although unsure if that was intentional).

Is the suggestion to not use multiple H1s? If so, I don't think it matters, but I can change it if you feel strongly about it.

Comment (rahimabdi) [reply accname/name/comp_text_node.html:19]: > Is the suggestion to not use multiple H1s? If so, I don't think it matters, but I can change it if you feel strongly about it.

Apologies, I didn't mean to be pedantic and definitely don't feel strongly about it in this context (have definitely had one too many spirited discussions about multiple `h1`s 😛?)

File: accname/name/comp_text_node.html (hunk 7 above)
Comment (rahimabdi) [reply accname/name/comp_text_node.html:14]: Created new AccName issue here: https://github.com/w3c/accname/issues/212

File: accname/name/comp_text_node.html (hunk 9 above)
Comment (cookiecrook) [reply accname/name/comp_text_node.html:16]: > * Text nodes as part of explicit ARIA roles

I will add this. Assuming this means `<div role="heading">` and `<div role="button" tabindex="0">` in addition to `<h2>` and `<button>`.

> * Tests that ensure labelledby/aria-label/name from content etc. take precedence over text nodes (may be captured in respective files)

I think the ones you wrote in #41463 test most of that, and more will come in https://github.com/web-platform-tests/interop-2023-accessibility-testing/issues/68 once unblocked.

Comment (cookiecrook) [reply accname/name/comp_text_node.html]: Resolved in https://github.com/web-platform-tests/wpt/pull/42407/commits/260d5e5354a86017d37439503660b90f85c1c856

File: accname/name/comp_text_node.html [hunk 10]
```diff
@@ -13,23 +13,139 @@
...
-</h2>
+  Note: some overlap with the tests in:
+  - /accname/name/comp_label.html
+  - /accname/name/comp_name_from_content.html
+
+-->
+
+<h1>text/element/text nodes, no space</h1>
+<span role="button" tabindex="0" class="ex" data-expectedlabel="buttonlabel" data-testname="span[role=button] with text/element/text nodes, no space">button<span></span>label</span>
+<div role="heading" class="ex" data-expectedlabel="headinglabel" data-testname="div[role=heading] with text/element/text nodes, no space">heading<span></span>label</div>
```
Comment (rahimabdi) [inline accname/name/comp_text_node.html:26]: The test looks good, somewhat related observation:

I noticed that HTML validation passes for this role="heading" without aria-level however, it's currently [a MUST that aria-level be provided](https://w3c.github.io/aria/#heading). Should this fail validation?

(This seems related to ARIA-only parsing errors like lack of explicit grouping for role="radio").

**Review by rahimabdi — APPROVED** [review]
Looks great!

File: accname/name/comp_text_node.html (hunk 10 above)
Comment (cookiecrook) [reply accname/name/comp_text_node.html:26]: It's an Author MUST, but it's I don't think it will change the results of these tests... If you know otherwise, we should have a new file testing how that and similar required attributes affect the label... Otherwise, heading level will get text after we have a direct accessor to it. (See https://github.com/WICG/aom/issues/197 and https://github.com/WICG/aom/issues/203)
